### PR TITLE
sample classifier pipeline and NCV tutorials + FLOWCHARTS

### LINKED ISSUE #318: ENH: q2-sample-classifier tutorial needs a flowchart
with an increasing number of actions, an expanded tutorial and flowchart are needed.

will add once pipelines are completed.

### LINKED ISSUE #329: sample-classifier tutorial: update to describe pipelines etc
- [x] describe new pipelines / outputs
- [x] CV vs. NCV
- [x] FLOWCHART
- ~step by step through individual actions? (to piece together pipelines)~

### PR DESCRIPTION
**Pull request by nbokulich**
fixes #329 
fixes #318 

~wait on https://github.com/qiime2/q2-sample-classifier/pull/120 before merging (that PR contains changes to some method/type names)~

### CODE REVIEW

File: source/tutorials/sample-classifier.rst [hunk 1]
```diff
@@ -39,21 +54,74 @@ Next, we will attempt to predict which body site each sample originated from bas
...
+
+.. command-block::
+
+   qiime metadata tabulate \
+     --m-input-file moving-pictures-classifier/feature_importance.qza \
+     --o-visualization moving-pictures-classifier/feature_importance.qzv
+
 
-If ``--p-optimize-feature-selection`` is enabled, the visualization will also display a recursive feature extraction plot, which illustrates how model accuracy changes as a function of feature count. The combination of features that maximize accuracy are automatically selected for the final model, which is used for sample prediction results that are displayed in the visualization. A list of the features chosen, and their relative importances, will be displayed at the bottom of the visualization. Features with higher importance scores are more important for distinguishing each class.
+If ``--p-optimize-feature-selection`` is enabled, only the selected features (i.e., the most important features, which maximize model accuracy) will be reported in this artifact. This allows us to not only see which features are most important, but also use that information to filter out uninformative features from our feature table:
```
Comment (thermokarst) [inline source/tutorials/sample-classifier.rst]: `--p-optimize-feature-selection` -> `--p-parameter-tuning`, right?

File: source/tutorials/sample-classifier.rst [hunk 2]
```diff
@@ -39,21 +54,74 @@ Next, we will attempt to predict which body site each sample originated from bas
...
+
+.. command-block::
+
+   qiime feature-table filter-features \
+     --i-table moving-pictures-table.qza \
+     --m-metadata-file moving-pictures-classifier/feature_importance.qza \
+     --o-filtered-table moving-pictures-classifier/important-feature-table.qza
+
+
+This pipeline also produces a visualization containing a summary of the model parameters used by the supervised learning estimator in ``model_summary.qzv``. If ``--p-optimize-feature-selection`` is enabled, the visualization will also display a recursive feature extraction plot, which illustrates how model accuracy changes as a function of feature count. The combination of features that maximize accuracy are automatically selected for the final model, which is used for sample prediction results that are displayed in the other outputs.
```
Comment (thermokarst) [inline source/tutorials/sample-classifier.rst]: Same as above regarding parameter name.

File: source/tutorials/sample-classifier.rst [hunk 3]
```diff
@@ -39,21 +54,74 @@ Next, we will attempt to predict which body site each sample originated from bas
...
 
 .. question::
    What happens when feature optimization is disabled with the option ``--p-no-optimize-feature-selection``? How does this impact classification accuracy?
```
Comment (thermokarst) [inline source/tutorials/sample-classifier.rst]: Does this param need to be updated? If so, can you do a find and replace?

File: source/tutorials/sample-classifier.rst (hunk 1 above)
Comment (nbokulich) [reply source/tutorials/sample-classifier.rst]: thanks for checking! `--p-optimize-feature-selection` is correct here.

File: source/tutorials/sample-classifier.rst (hunk 2 above)
Comment (nbokulich) [reply source/tutorials/sample-classifier.rst]: correct here

File: source/tutorials/sample-classifier.rst [hunk 4]
```diff
@@ -28,7 +28,22 @@ Supervised learning classifiers predict the categorical metadata classes of unla
    :url: https://data.qiime2.org/2018.8/tutorials/sample-classifier/moving-pictures-table.qza
    :saveas: moving-pictures-table.qza
 
-Next, we will attempt to predict which body site each sample originated from based on its microbial composition.
+Next, we will train and test a classifier that predicts which body site a sample originated from based on its microbial composition. We will do so using the ``classify-samples`` pipeline, which performs a series of steps under the hood:
+
+1. The input samples are randomly split into a ``training`` set and a ``test`` set. The test set is held out until the end of the pipeline, allowing us to test accuracy on a set of samples that was not used for model training. The fraction of test samples to remove is adjusted with the ``--p-test-size`` parameter.
```
Comment (gregcaporaso) [inline source/tutorials/sample-classifier.rst]: The word _remove_ doesn't seem right here: _The fraction of test samples to remove is adjusted with the_

How about: _The fraction of input samples to include in the test set is adjusted with the_

File: source/tutorials/sample-classifier.rst [hunk 5]
```diff
@@ -39,21 +54,74 @@ Next, we will attempt to predict which body site each sample originated from bas
      --p-optimize-feature-selection \
      --p-parameter-tuning \
      --p-estimator RandomForestClassifier \
-     --p-n-estimators 100 \
+     --p-n-estimators 20 \
      --output-dir moving-pictures-classifier
 
-The visualization produced by this command presents classification accuracy results in the form of a confusion matrix. This matrix indicates how frequently a sample is classified with to the correct class vs. all other classes. The confusion matrix is displayed at the top of the visualization in the form of a heatmap, and below that as a table containing overall accuracy (the fraction of times that test samples are assigned the correct class).
+
+This pipeline produces several outputs. First let's check out ``accuracy_results.qzv``, which presents classification accuracy results in the form of a confusion matrix. This matrix indicates how frequently a sample is classified with to the correct class vs. all other classes. The confusion matrix is displayed at the top of the visualization in the form of a heatmap, and below that as a table containing overall accuracy (the fraction of times that test samples are assigned the correct class).
```
Comment (gregcaporaso) [inline source/tutorials/sample-classifier.rst]: _a sample is classified with to the correct class _ -> _a sample is classified with the correct class _

File: source/tutorials/sample-classifier.rst [hunk 6]
```diff
@@ -39,21 +54,74 @@ Next, we will attempt to predict which body site each sample originated from bas
...
+
+.. command-block::
+
+   qiime metadata tabulate \
+     --m-input-file moving-pictures-classifier/predictions.qza \
+     --o-visualization moving-pictures-classifier/predictions.qzv
+
+
+Another really useful output of supervised learning methods is *feature selection*, i.e., they report which features (e.g., ASVs or taxa) are most predictive. A list of all features, and their relative importances, will be reported in ``feature_importance.qza``. Features with higher importance scores are more important for distinguishing each class. Note that some estimators — notably K-nearest neighbors models — do not report feature importance scores, so this output will be meaningless if you are using such an estimator. Feature importances are of the semantic type ``FeatureData[Importance]``, and can be interpreted as (feature) metadata so we can take a look at these feature importances (and/or :ref:`merge with other feature metadata <exploring feature metadata>`) using ``metadata tabulate``:
```
Comment (gregcaporaso) [inline source/tutorials/sample-classifier.rst]: _are more important for distinguishing each class_ -> _were more useful for distinguishing classes_

File: source/tutorials/sample-classifier.rst [hunk 7]
```diff
@@ -39,21 +54,74 @@ Next, we will attempt to predict which body site each sample originated from bas
...
    What happens when feature optimization is disabled with the option ``--p-no-optimize-feature-selection``? How does this impact classification accuracy?
 
-K-fold cross-validation is performed during automatic feature selection and parameter optimization steps. Five-fold cross-validation is performed by default, and this value can be adjusted using the ``--p-cv`` parameter. A separate portion of samples is removed from the data set prior to model training and optimization, and used as a test set to determine model accuracy. The fraction of test samples to remove is adjusted with the ``--p-test-size`` parameter.
+Finally, the trained classification model is saved down for convenient re-use in the ``sample_estimator.qza`` artifact! This allows us to predict metadata values for additional samples. For example, imagine we just received a shipment of new samples and wanted to use our pre-trained Body Site classifier to figure out what type of samples these new samples are. For the sake of convenience in this example, we will just pretend we have new samples and predict the values of the same samples that we used to train the model but **NEVER do this in practice** because:
```
Comment (gregcaporaso) [inline source/tutorials/sample-classifier.rst]: _saved down_ -> _saved_

**Comment by thermokarst** [comment]
@gregcaporaso, can you open a new issue to link to these proposed changes, since this PR was already merged two weeks ago? Thanks!

**Pull request merged**